lib/usecases/environment/GetAllEnvironmentsUseCase.js

```diff
@@ -58,14 +58,26 @@ class GetAllEnvironmentsUseCase {
             }
         }
 
-        const { count, rows } = await TransactionHelper.provide(async () => {
+        let { count, rows } = await TransactionHelper.provide(async () => {
             queryBuilder.orderBy('updatedAt', 'desc');
             queryBuilder.limit(limit);
             queryBuilder.offset(offset);
             queryBuilder.include({ association: 'runs', separate: true });
             queryBuilder.include({ association: 'historyItems', separate: true });
             return EnvironmentRepository.findAndCountAll(queryBuilder);
         });
+
+        // Apply these filters after getting all environments to avoid unnecessary code for getting environment status history
+        if (filter?.currentStatus) {
+            const filters = filter.currentStatus.split(',').map((id) => id.trim());
+            rows = rows.filter((environment) => {
+                const { historyItems } = environment;
+                const lastHistoryItem = historyItems?.length > 0 ? historyItems[historyItems.length - 1] : null;
+                return filters.includes(lastHistoryItem?.status);
+            });
+            count = rows.length;
+        }
```

Comment:
Because it is an association, you will indeed have to do the querry in two steps: first fetch the environment by filtering them, and then fetch the associations.
This way, you will probably not need `separate: true` btw

The filtering have to be done in sequelize query, because it will be much quicker